### PR TITLE
Add userlist command

### PR DESCRIPTION
**Pull request by panpawn**
A lot of users have been asking for a userlists command, may it be for bot purposes or for battle room usage.

PREVIEW:

### CODE REVIEW

**Comment by Zarel** [comment]
Which users? What do they say?

**Comment by panpawn** [comment]
I've seen people be like "I want a way to see a list of users who are in my battle"

I've also seen people request it for their bots.

I think this might be a good way to hit both of those usages, although I do understand that a client userlist for battles would be more appropriate for that half of it's usage.  With that in mind, maybe this is a good solution for that until that becomes a thing?

**Comment by acz13** [comment]
If this is broadcast shouldn't it be limited to a certain # of people?

Could be spammy in rooms with lots of people...

**Comment by panpawn** [comment]
Actually, in thinking about it, I see no reason why this should be broadcast-able at all.

File: chat-commands.js [hunk 1]
```diff
@@ -70,6 +70,24 @@ exports.commands = {
...
+	userlist: function (target, room, user) {
+		let userList = [];
+
+		for (let i in room.users) {
+			let curUser = Users(room.users[i]);
+			if (!curUser || !curUser.named) continue;
+			userList.push(Chat.escapeHTML(curUser.name));
+		}
+
+		let output = "There " + Chat.plural(userList.length, 'are', 'is') + " <strong style=\"color:#24678d\">" + userList.length + "</strong> user" + Chat.plural(userList.length, 's', '') + " in this room:<br />";
```
Comment (Zarel) [inline chat-commands.js]: Template string...

File: chat-commands.js [hunk 2]
```diff
@@ -70,6 +70,24 @@ exports.commands = {
...
+
+		userList = userList.sort(); // sort names alphabetically
+		output += userList.join(', ');
```
Comment (Zarel) [inline chat-commands.js]: Template string...

File: chat-commands.js (hunk 1 above)
Comment (Zarel) [inline chat-commands.js]: Also `Chat.plural(userList.length, 's', '')` is just `Chat.plural(userList.length)`

Comment (panpawn) [inline chat-commands.js]: Done and done

File: chat-commands.js (hunk 2 above)
Comment (panpawn) [inline chat-commands.js]: Done

**Comment by panpawn** [comment]
Updated

File: chat-commands.js [hunk 3]
```diff
@@ -70,6 +70,24 @@ exports.commands = {
 		"/auth [room] - Show what roomauth a room has.",
 		"/auth [user] - Show what global and roomauth a user has."],
 
+	userlist: function (target, room, user) {
+		let userList = [];
+
+		for (let i in room.users) {
+			let curUser = Users(room.users[i]);
+			if (!curUser || !curUser.named) continue;
+			userList.push(Chat.escapeHTML(curUser.name));
```
Comment (Zarel) [inline chat-commands.js]: Should probably include rank, especially if bots are going to be using this.

Comment (Zarel) [inline chat-commands.js]: In fact maybe just use the existing userlist getter...

Comment (panpawn) [inline chat-commands.js]: Done

**Comment by panpawn** [comment]
Updated

File: chat-commands.js (hunk 3 above)
Comment (panpawn) [inline chat-commands.js]: er, not done :ss

didn't see second comment, updating now

Comment (panpawn) [inline chat-commands.js]: ok, now done

**Comment by panpawn** [comment]
updated v3

File: chat-commands.js [hunk 4]
```diff
@@ -70,6 +70,16 @@ exports.commands = {
 		"/auth [room] - Show what roomauth a room has.",
 		"/auth [user] - Show what global and roomauth a user has."],
 
+	userlist: function (target, room, user) {
+		let userlist = room.getUserList().substr(7);
+		userlist = userlist.split(',').splice(1); // split userlist by comma, exclude usercount
```
Comment (Zarel) [inline chat-commands.js]: > `splice`

...???????

`splice` is a very unperformant (you shouldn't ever use it unless you need to mutate an array for whatever reason, usually `slice`/`concat` is better) and isn't even doing anything here (`splice` needs at least two arguments)

Comment (Zarel) [inline chat-commands.js]: In hindsight, `getUserList` is probably wrong; implementing it yourself like you did before seems to be the best choice.

Comment (panpawn) [inline chat-commands.js]: Good to know, thank you

**Comment by panpawn** [comment]
ok, I think i did the thingy

File: chat-commands.js (hunk 4 above)
Comment (panpawn) [inline chat-commands.js]: >and isn't even doing anything here

``>> ['this', 'is', 'a', 'test'].splice(1);``
``<< is,a,test``

**Comment by Morfent** [comment]
I'm late, but I have a couple of concerns with the design of this:

1. HTML is the least friendly way to design this if it's for bots in mind since it requires wrangling the string the bot receives and the way the string's formatted may or may not even remain the some from now onward
  - Bots can already keep track of userlists on their own anyways given `|j|`/`|J|`/`|l|`/`|L|`/`|N|` messages, as well as the `|users|` message when joining regular chatrooms, regardless
2. If we need a command to see the userlist in battle rooms, isn't that more of an issue with the client's battle room design?  This seems like a bandaid over the real issue, which the average user likely wouldn't be aware of outside of word of mouth/reading the code

If we really need a way for bots to grab the userlist at any time, I prefer https://github.com/Zarel/Pokemon-Showdown/pull/2582 as a solution

**Pull request merged**